Two misc ideas for Nix itself (#13)
NixOS/GSoC: ideas/2024.md

## small (90 hours)

### Nix cleanup: merge `ParsedDerivation` into `BasicDerivation`, improve JSON format

Nix currently special-cases some environment variables in derivations to convey extra meaning ("system features", "prefer local build", etc.).

> We're stuck with this in the `.drv` A-Term format and `builtins.derivation` because backwards compatibility, but that doesn't mean our internal data structures and (currently experimental, i.e. not yet stabilized and immobile) JSON format cannot be improved.

Reviewer: what is a A-Term?

Author: ```suggestion
We're stuck with this in the [`.drv` [A-Term format](https://nixos.org/manual/nix/stable/protocols/derivation-aterm) and `builtins.derivation` because backwards compatibility, but that doesn't mean our internal data structures and (currently experimental, i.e. not yet stabilized and immobile) JSON format cannot be improved.
```

> We're stuck with this in the [`.drv` A-Term format](https://nixos.org/manual/nix/stable/protocols/derivation-aterm) and `builtins.derivation` because backwards compatibility, but that doesn't mean our internal data structures and (currently experimental, i.e. not yet stabilized and immobile) JSON format cannot be improved.

Reviewer 2: ```suggestion
We're stuck with this in the [`.drv` A-Term format](https://nixos.org/manual/nix/stable/protocols/derivation-aterm) and `builtins.derivation` because of backwards compatibility requirements, but that doesn't mean our internal data structures and (currently experimental, i.e. not yet stabilized and immobile) JSON format cannot be improved.
```

## medium (175 hours)

### Nix Internals: Use `std::filesystem::path` for `Path`

Nix currently has a bunch of home-grown file system and file path utilities.

> However in recently years (C++2017) the standard library has gained a new `std::fileystem` library which has much of the same functionality.

Author: Err I mean https://en.wikipedia.org/wiki/C%2B%2B17; the year of the standard usually isn't actually the year it appeared.

> However in recently years (C++2017) the standard library has gained a new `std::fileystem` library which has much of the same functionality.

Author: ```suggestion
However in recently years ([C++17](https://en.wikipedia.org/wiki/C%2B%2B17#Language)) the standard library has gained a new `std::fileystem` library which has much of the same functionality.
```

> However in recently years (C++2017) the standard library has gained a new `std::fileystem` library which has much of the same functionality.

Reviewer: I'm still convinced that it should be in recent years and not recently, but my brain is currently very mush.

> However in recently years (C++2017) the standard library has gained a new `std::fileystem` library which has much of the same functionality.

Author: yes good point re typo, added that to the suggestion above.
